herd/ASLSem.ml

```diff
@@ -160,8 +161,35 @@ module Make (C : Config) = struct
       | 128 -> MachSize.S128
       | _ -> Warn.fatal "Cannot access a register with size %s" (V.pp_v v)
 
-    let to_bv = M.op1 (Op.ArchOp1 ASLValue.ASLArchOp.ToBV)
-    let to_int = M.op1 (Op.ArchOp1 ASLValue.ASLArchOp.ToInt)
+    let to_bv v =
+      match v with
+      | V.Val (Constant.Concrete (ASLScalar.S_BitVector _)) -> return v
+      | V.Var _ | V.Val (Constant.Concrete _) ->
+          M.op1 (Op.ArchOp1 ASLValue.ToBV) v
+      | v ->
+          let v' = V.fresh_var () in
+          M.restrict M.VC.[ Assign (v', Unop (Op.ArchOp1 ASLValue.ToBV, v)) ]
+          >>! v'
+
+    let to_int_unsigned v =
+      match v with
+      | V.Val (Constant.Concrete (ASLScalar.S_Int _)) -> return v
+      | V.Var _ | V.Val (Constant.Concrete _) ->
+          M.op1 (Op.ArchOp1 ASLValue.ToIntU) v
+      | v ->
+          let v' = V.fresh_var () in
+          M.restrict M.VC.[ Assign (v', Unop (Op.ArchOp1 ASLValue.ToIntU, v)) ]
+          >>! v'
+
+    let to_int_signed v =
+      match v with
+      | V.Val (Constant.Concrete (ASLScalar.S_Int _)) -> return v
+      | V.Var _ | V.Val (Constant.Concrete _) ->
+          M.op1 (Op.ArchOp1 ASLValue.ToIntS) v
+      | v ->
+          let v' = V.fresh_var () in
+          M.restrict M.VC.[ Assign (v', Unop (Op.ArchOp1 ASLValue.ToIntS, v)) ]
+          >>! v'
```

Comment:
Same remark calling `M.op1` may do the job or .... ?

Reply:
Done in cac5b6aa5659f8ccdb3597f39b96c99099c4ea1e